### ISSUE TITLE
Nix 2.0.1: Assertion `info.ca == "" || info.isContentAddressed(*this)' failed: fixed-output derivations can contains references.

### ISSUE THREAD

**Issue by hvdijk** [NixOS/nix#2133]
Reduced self-contained test case:

    nix-build --no-out-link --expr '
        with builtins;
        with import <nix/config.nix>;
        rec {
            test-1 = derivation {
                name = "test-1";
                system = currentSystem;
                builder = shell;
                args = ["-c" ">$out"];
            };
            test-2 = derivation {
                name = "test-2";
                outputHash = hashString "sha256" "${test-1}";
                outputHashAlgo = "sha256";
                system = currentSystem;
                builder = shell;
                args = ["-c" "echo -n ${test-1} >$out"];
        };
    }'

This worked in 2.0, but in 2.0.1 produces:

    these derivations will be built:
      /nix/store/4rj1xzkd63mll3d6n0886gw4fmqxy15s-test-1.drv
      /nix/store/130xbgj5p2xdhkp0nrjx2pppjbq5jl9c-test-2.drv
    building '/nix/store/4rj1xzkd63mll3d6n0886gw4fmqxy15s-test-1.drv'...
    building '/nix/store/130xbgj5p2xdhkp0nrjx2pppjbq5jl9c-test-2.drv'...
    warning: path '/nix/store/gkm7n9hf4g6d8iz3vncpkim7wdbq5s3v-test-2' claims to be content-addressed but isn't
    nix-build: src/libstore/local-store.cc:584: uint64_t nix::LocalStore::addValidPath(nix::LocalStore::State&, const nix::ValidPathInfo&, bool): Assertion `info.ca == "" || info.isContentAddressed(*this)' failed.
    Aborted

This is caused by commit 247630287b09544056461c98fc1f0d5580e4c368, which asserts that fixed-output derivations cannot contain references, but as this test case shows, that assertion is incorrect.

**Comment by copumpkin** [comment]
Oh, I've been getting a lot of that too, but it's not consistent! Thanks for tracking this down!

**Comment by edolstra** [comment]
I've changed the assertion failure into an exception. The check is correct however. Content-addressable paths currently cannot have references because the references are not taken into account when computing the store path.

https://github.com/edolstra/nix/commit/236e87c687065c2aa935542533b1a65282a73347 adds support for references in content-addressable paths, but it's not on master yet.

**Comment by hvdijk** [comment]
Sorry, I don't understand. Is it correct and intended for the 2.0 -> 2.0.1 update to break working derivations and render Nix unusable for my system? Does this mean I should just stick with 2.0?

I understand that the references weren't taken into account when computing the store path. They don't need to be, at least not here, because the store path for fixed-output derivations gets determined entirely by the name and output hash, not by any references to other derivations.

**Comment by copumpkin** [comment]
Yeah, also curious about this, because it affects a lot of what I'm doing and it seems weird to break nontrivial stuff in a minor update. Maybe the error message can point at what to do if we encounter it? It's not currently very helpful

**Comment by hvdijk** [comment]
Thanks for the extra commit. Testing 2.0.1 + your commits now, can confirm it's working again.

**Comment by copumpkin** [comment]
@edolstra I think I'm seeing this intermittently still in 2.0.2, but haven't been able to repro it consistently yet 😦

**Comment by copumpkin** [comment]
@edolstra would you expect `nix-store --query --requisites` to only contain one entry on any fixed-output derivation path in the store? The path I see it on intermittently has more than one.

**Comment by copumpkin** [comment]
Just to elaborate a bit on what I'm seeing and what might be causing it:

I have this .narinfo in my binary cache on the derivation that seems to fail intermittently (still can't get it to fail consistently)

```
StorePath: /nix/store/8w9r8q2pvcil6v4syjyrp16lj0xc7hzf-source
URL: nar/17hwp05kdzk2i6jz04hn0k9zy4p77j2vs9r5kr5d79klh0arjhk2.nar.xz
Compression: xz
FileHash: sha256:17hwp05kdzk2i6jz04hn0k9zy4p77j2vs9r5kr5d79klh0arjhk2
FileSize: 9402776
NarHash: sha256:0y3jrpzas6n1cawjn2gp4mhz9r45x5kb29gq4447jx0qhm18hhan
NarSize: 54661848
References: kvfaf1ayfqgg26q37drg768q2wcqjdqh-other-source
Deriver: dmia59amnczszdgrp0srjpzhliir6r4d-source.drv
CA: fixed:r:sha256:0y3jrpzas6n1cawjn2gp4mhz9r45x5kb29gq4447jx0qhm18hhan
```

Note how it's both `fixed:` and has an entry in `References`, which as far as I understand, the content-addressable check dislikes.

**Comment by copumpkin** [comment]
Aha, got it consistently on 2.0.2. The intermittent behavior was due to `~/.cache/nix/binary-cache-v5.sqlite` and when I wipe that between runs I can get it to run consistently. Putting together a minimal repro now.

**Comment by copumpkin** [comment]
Okay, I've reduced this to a minimal example that fails this way on 2.0.2, but requires a bit of hackery to reproduce. Start with a default.nix containing the following:

```nix
rec {
  foo = builtins.path {
    name = "foo";
    path = ./foo;
  };

  bar = (import <nixpkgs> {}).runCommand "bar" {
    outputHash     = "09c3n24xyj8ab8hsxbk0lfww4d2fpkfss4qh4yhy7zdcyp791y9q";
    outputHashMode = "recursive";
    outputHashAlgo = "sha256";
  } ''
    mkdir -p $out/foo
    cp -r ${foo} $out/foo
    echo "bar" > $out/bar
  '';
}
```

(foo/blah is an empty file and foo/quux contains the string "foobar\n", or you can just change the `outputHash` if you don't want to try to replicate that)

Now `nix-build -A bar` in that folder. It should produce something like `/nix/store/x30qynnir0i46qd5gq2mff6jxlvrb3xn-bar` on Linux. No content-addressable errors yet (I think they would've appeared in 2.0.1 though).

Now make a folder for a local binary cache:

```shell
mkdir ~/binary-cache
nix copy --to file:///path/to/binary-cache $(nix-build -A bar)
```

You should end up with two entries in that binary cache, one for `foo` and the other for `bar`. If you check the .narinfo for `bar`, it'll mention `foo` as a reference. In my 2.0.2, this .narinfo didn't include the `CA` field. This is important.

So now delete `foo` and `bar` from your store. `nix-store -r --option binary-caches file:///path/to/binary-cache` your `bar` store path back to existence. This should work fine too, and should bring both `foo` and `bar` back in from the binary cache.

Now for the bug: add the `CA` field back in to the .narinfo, since previous versions of Nix seemed to put it there (so my binary cache is full of it). It's easy to compute, just take the `NarHash` and stick `fixed:r:` in front of it. Now delete `~/.cache/nix/binary-cache-v5.sqlite` and `nix-store -r` the path back in. Bam, you should get the bug:

```shell
# nix-store -r /nix/store/x30qynnir0i46qd5gq2mff6jxlvrb3xn-bar --option binary-caches file:///root/binary-cache/
these paths will be fetched (0.00 MiB download, 0.00 MiB unpacked):
  /nix/store/jn5gxfvlk23dq1hkrcwwys289ygdc2kz-foo
  /nix/store/x30qynnir0i46qd5gq2mff6jxlvrb3xn-bar
copying path '/nix/store/jn5gxfvlk23dq1hkrcwwys289ygdc2kz-foo' from 'file:///root/binary-cache/'...
copying path '/nix/store/x30qynnir0i46qd5gq2mff6jxlvrb3xn-bar' from 'file:///root/binary-cache/'...
warning: path '/nix/store/x30qynnir0i46qd5gq2mff6jxlvrb3xn-bar' claims to be content-addressed but isn't
cannot add path '/nix/store/x30qynnir0i46qd5gq2mff6jxlvrb3xn-bar' to the Nix store because it claims to be content-addressed but isn't
error: build of '/nix/store/x30qynnir0i46qd5gq2mff6jxlvrb3xn-bar' failed
```

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
